components/server/src/iam/iam-session-app.ts

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { injectable, inject } from "inversify";
+import * as express from "express";
+import { IamLoginHandler } from "./iam-login-handler";
+import { SessionHandlerProvider } from "../session-handler";
+import { Authenticator } from "../auth/authenticator";
+
+@injectable()
+export class IamSessionApp {
+    @inject(IamLoginHandler)
+    protected readonly loginHandler: IamLoginHandler;
```

Comment:
Why the indirection and injection here when we only have a single place that uses it?

Reply:
The decomposition improves testability. 
Injection is used to saturate dependencies.

Reply:
> Injection is used to saturate dependencies.

What does that mean?

Reply:
`IamLoginHandler` depends on `UserService`, so the whole dependency tree will be resolved using DI. That's how configuration is managed in server.

Reply:
But how is that *used to saturate dependencies*? I don't understand what that means. What is there to saturate when they are all just singletons?

Sure, my question was more that the only place where we need this handler is in this file. It feels that we could just in-line it and depend on the user service in this class, rather than create the extra level of indirection.

Reply:
In general, we're using DI instead for configuration management. So, no `newService(new Provider())`, instead everything is bound in a DI container to be injected as required. 
In case of tests, what we do, we create test containers to provide the dependencies for the isolated unit to test.

> It feels that we could just in-line it and depend on the user service in this class, rather than create the extra level of indirection.

Ah, in case you're talking about inlining the handler into the app: the handler itself is much easier to test in isolation, without providing all of the dependencies for the `IamSessionApp`. For instance, all of the DB-based store for session is pulled in with the `sessionHandlerProvider.sessionHandler` middleware. That's no longer a plain unit test then. In test, I'd go with a plain session store from express, which would work with no further changes.